peer/waiting_handler.go

```diff
@@ -34,5 +34,10 @@ func (w *waitingResponseHandler) OnFailure() error {
 
 // newWaitingResponseHandler returns new instance of the waitingResponseHandler
 func newWaitingResponseHandler() *waitingResponseHandler {
-	return &waitingResponseHandler{responseChan: make(chan []byte)}
+	return &waitingResponseHandler{
+		// Make buffer length 1 so that OnResponse can complete
+		// even if no goroutine is waiting on the channel (i.e.
+		// the context of a request is cancelled.)
+		responseChan: make(chan []byte, 1),
```

Comment:
Can we add a regression test to catch the error where the response handlers deadlock if we did not initialize the responseChan to size 1?

Reply:
Added for `TestAppRequestAnyOnCtxCancellation`. I can also add similar logic to other tests if you want -- lmk.